Implementation of tabu list

ZedongPeng opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

ZedongPeng committed:
half-way tuba list

ZedongPeng committed:
Merge branch 'mindtpy-master' into feasibility_pump

ZedongPeng committed:
Merge branch 'mindtpy-master' into feasibility_pump

ZedongPeng committed:
half way tabu list

ZedongPeng committed:
fix typo

ZedongPeng committed:
no bug

ZedongPeng committed:
update use_tabu_list description

ZedongPeng committed:
tabu_list almost done

bernalde commented on pyomo/contrib/mindtpy/MindtPy.py:
```diff
@@ -95,7 +96,9 @@ def solve(self, model, **kwds):
 
         if config.strategy == "GOA":
-            config.add_no_good_cuts = True
+            # TODO: Choose one from the following two
```
This is done based on numerical results.

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -126,6 +127,18 @@ def solve_master(solve_data, config, feas_pump=False):
+        masteropt._solver_model.parameters.preprocessing.reduce.set(1)
+        # If the callback is used to reject incumbents, the user must set the 
+        # parameter c.parameters.preprocessing.reduce either to the value 1 (one) 
+        # to restrict presolve to primal reductions only or to 0 (zero) to disable all presolve reductions
```
Does this mean that those extra parameters should be fixed by us or rely on the final user to do it? Is this note for us the developers?

ZedongPeng replied:
This is the note for us and it's from the CPLEX website. I test it.
In cplex, if the incumbent solution (optimal) is found at the root node(before branch & bound), the incumbent callback will not be called. This often happens in small MIPs.
If we set this parameter to 1, it will prevent the above situation from happening.

bernalde replied:
This was the same note in the paper. I think they handled it by actually adding the no-good cut in that case. I don't want to turn off the presolve.

ZedongPeng replied:
Actually, I am not turning off the whole presolve. I just turn off dual reduction and the primal reduction still applies. 

Incumbent callback reference:
https://www.ibm.com/support/knowledgecenter/en/SSSA5P_12.7.1/ilog.odms.cplex.help/refpythoncplex/html/cplex.callbacks.IncumbentCallback-class.html

Preprocessing reference:
https://www.ibm.com/support/knowledgecenter/SSSA5P_12.9.0/ilog.odms.cplex.help/CPLEX/Parameters/topics/Reduce.html

ZedongPeng replied:
As you can see in the document, the dual reduction will be automatically turned off if `lazyconstraint` callback is included.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -9,6 +9,8 @@
 from pyomo.core.expr import current as EXPR
 from pyomo.opt import SolverFactory
 from pyomo.solvers.plugins.solvers.persistent_solver import PersistentSolver
+import cplex
```
These imports should be made through _is_available. Currently, tests fail because CPLEX might not be available.

ZedongPeng replied:
I think we can refer to here.
https://github.com/Pyomo/pyomo/pull/1391#discussion_r415933190

Should I create a new file to place the incumbent callback?

bernalde replied:
Yes, that should solve the problem.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -318,3 +320,22 @@ def generate_norm1_norm_constraint(model, setpoint_model, config, discrete_only=
+            integer_var_value_list.append(round(self.get_values(opt._pyomo_var_to_solver_var_map[var])))
+        integer_var_value_tuple = tuple(integer_var_value_list)
+        if integer_var_value_tuple in solve_data.tabu_list:
+            self.reject()
```
This looks WAY simpler than what I thought. I think one key issue might be performance with large problems, so there should be an option to write a temporary file with those solutions.

ZedongPeng replied:
Actually, I thought about this. So I use `set()` instead of `list()` here. I don't see the point to write a temporary file since we still need to read it, check if it's in the `list` or `set`, then write it. 
Do you mean writing the final tabu_list?

bernalde replied:
I think that having the last tabu list might be useful (at least printing it in DEBUG logging) and that should solve the issue. The good part about writing the file is that you don't need to keep in memory `solve_data.tabu_list`, but you are right, we can live with that.

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_global.py:
```diff
@@ -125,10 +125,7 @@ def test_GOA_MINLP2_simple(self):
                                 init_strategy='initial_binary',
                                 mip_solver=required_solvers[1],
-                                nlp_solver=required_solvers[0],
-                                obj_bound=10,
```
The objective bound is no longer required?

bernalde commented on pyomo/contrib/mindtpy/single_tree.py:
```diff
@@ -472,7 +472,7 @@ def handle_lazy_subproblem_other_termination(self, fixed_nlp, termination_condit
         """
         This is an inherent function in LazyConstraintCallback in cplex. 
-        This funtion is call whenever the a integer solution is found during the branch and bound process
+        This funtion is called whenever the a integer solution is found during the branch and bound process
```
function typo

bernalde reviewed (commented):
This ended up being simpler than I expected. Does it work? Is it faster than the cuts? What happens with problems with integer (not binary) variables?

ZedongPeng committed:
fix typo

ZedongPeng committed:
add check_config function

ZedongPeng committed:
move tabu_list to a new file

bernalde commented on pyomo/contrib/mindtpy/tabu_list.py:
```diff
@@ -0,0 +1,19 @@
+from cplex.callbacks import IncumbentCallback
```
In this file is where you are doing the cplex_available import. The tests are still failing because we import cplex directly and some testing computers don't have access to it.

ZedongPeng committed:
change tabu_list datatype from set to list, add tabu_list in bound fix, fix obj error

ZedongPeng commented:
Hey David @bernalde . For the `tabulist`, I defined it as `set()` in Python, which would be more efficient if we want to check if the current solution is in it. However, in bound_fix, we need to deactivate the cuts (here the `tabulist`) that we don’t trust. There is no order in `set()` that we can refer to. So I changed it to `list`.
I also fixed some bug.
